[design] 프로필 이미지 및 닉네임 입력 화면 생성

green-yoon87 opened the pull request:
### 🔖 Issue Number

https://github.com/boostcampwm-2024/iOS03-SniffMeet/issues/24#issue-2645682339


### 📙 작업 내역

1. 피그마에 맞게 프로필 및 이미지 생성 뷰 UI를 만들었습니다. 

| 경고 라벨 활성화 ver |  경고 라벨 비활성화 ver |
| --- | --- |
|<img width="240" alt="스크린샷 2024-11-07 오후 3 18 08" src="https://github.com/user-attachments/assets/41a0b886-f127-497a-ba6b-ddeee468e80e">  | <img width="240" alt="스크린샷 2024-11-07 오후 3 18 08" src="https://github.com/user-attachments/assets/a8d090ac-d80c-4287-b210-0a2afffb2ac7"> |

2. PHPickerViewcontroller를 이용하여 갤러리에서 이미지를 가져올 수 있도록 구현했습니다. 

```swift 
 func picker(_ picker: PHPickerViewController, didFinishPicking results: [PHPickerResult]) {
      ...
         itemProvider.loadObject(ofClass: UIImage.self) { image, error in
             guard let selectedImage = image as? UIImage else { return }
             DispatchQueue.main.async { [weak self] in
                 self?.profileImageView.image =  selectedImage
             }
         }
    }
```
main thread가 아닌 다른 thread에서 해당 코드가 실행될 위험이 있어서 
imageview의 image를 바꿀 때 mainThread에서 변경할 수 있도록 이렇게 코드를 작성했는데요. 제가 맞게 판단한건지 팀원분들의 의견이 궁금합니다. 

3. TextField 텍스트의 유무에 따른 버튼 활성화 

2, 3번 실행 화면 
Simulator Screen Recording - iPhone 16 - 2024-11-09 at 23.58.13.mp4

4. Common 폴더의 PrimaryButton에 버튼의 상태에 따라 backgroundColor 설정 
- 해당 버튼에 버튼 상태에 따라 배경색을 설정하는 코드가 빠져있어 추가로 작성했습니다. 

### 📝 PR 특이사항 
- Viper 패턴에 따르면 Viewcontroller 파일이 View 폴더에 속합니다. 따라서 view controller 파일을 생성할 때 ~view라고 명명하면 어떨지 제안드리고 싶습니다! 
- 프로필 및 닉네임 설정 화면이어서 이름을 ProfileSetupView라고 지었는데요. 이름이 어색하지 않은지 궁금합니다... 의견 주시면 수정해볼게요!

green-yoon87 committed:
[design] #24 UIView를 동그랗게 만드는 함수 extension에 추가

green-yoon87 committed:
[design] #24 닉네임 및 프로필 설정 화면 뷰 만들기

green-yoon87 committed:
[fix] #24 print문 제거

green-yoon87 committed:
[feat] #24 configuration 이용하여 버튼의 상태에 따라 다른 배경색 적용할 수 있도록 추가

green-yoon87 committed:
[feat] #24 PHPickerController 이용하여 이미지 가져오기 및 텍스트 필드의 텍스트에 따른 버튼 활성화 설정

textfield에 한글자 이상이라도 적혀있으면 버튼을 활성화할 수 있도록 설정했습니다.
또한, PHPickerViewController를 이용하여 갤러리에서 이미지를 가져올 수 있도록 설정했습니다.

soletree commented on SniffMeet/SniffMeet.xcodeproj/project.pbxproj:
```diff
@@ -7,21 +7,35 @@
-		3200437C2CDCA6F300D08B6D /* TestViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 3200437B2CDCA6F300D08B6D /* TestViewController.swift */; };
-		A2C844D12CDBE1E5007F2970 /* PrimaryButton.swift in Sources */ = {isa = PBXBuildFile; fileRef = A2C844D02CDBE1E5007F2970 /* PrimaryButton.swift */; };
-		A2C844D72CDBEF8B007F2970 /* KeywordButton.swift in Sources */ = {isa = PBXBuildFile; fileRef = A2C844D62CDBEF8B007F2970 /* KeywordButton.swift */; };
+		320043702CDC9A0D00D08B6D /* (null) in Sources */ = {isa = PBXBuildFile; };
+		320043722CDC9E5F00D08B6D /* (null) in Sources */ = {isa = PBXBuildFile; };
+		320043762CDCA18E00D08B6D /* (null) in Sources */ = {isa = PBXBuildFile; };
+		320043782CDCA49100D08B6D /* (null) in Sources */ = {isa = PBXBuildFile; };
+		3200437C2CDCA6F300D08B6D /* (null) in Sources */ = {isa = PBXBuildFile; };
```
P5. 이거... null 로 잡히는거는 뭐뭣일까요...

green-yoon87 replied:
Common 폴더에 있는 파일에 로컬에 있는데도 불구하고 Xcode에 보이지 않는게 조심스레 이 부분 때문이라고 생각이 듭니당...

soletree replied:
요상하군요🫠 이 부분 이슈로 따로 빼서 해결을 해봐야겠네요...

soletree commented on SniffMeet/SniffMeet/Source/Auth/CreateProfile/View/ProfileSetupView.swift:
```diff
@@ -0,0 +1,156 @@
+
+extension ProfileSetupView: UITextFieldDelegate {
+    func textFieldDidChangeSelection(_ textField: UITextField) {
+        if (textField.text?.count ?? 0 < 1) || (textField.text?.count ?? 0 < 1) {
+            submitButton.isEnabled = false
+        } else {
+            submitButton.isEnabled = true
+        }
```
P2. 요거 동일한 조건문이 OR로 연결되어 있는거 맞을까요? 
P4. `submitButton.isEnabled = (textField.text?.count ?? 0 > 1)` 요렇게도 써볼 수 있을 것 같습니다! 지금 코드도 좋아요

green-yoon87 replied:
OR로 연결되어 있는건 수정이 필요한거 같네요 
추천해주신 코드 방식 간결해서 너무 좋습니당!

soletree commented on SniffMeet/SniffMeet/Source/Auth/CreateProfile/View/ProfileSetupView.swift:
```diff
@@ -0,0 +1,156 @@
+           itemProvider.canLoadObject(ofClass: UIImage.self) {
+            itemProvider.loadObject(ofClass: UIImage.self) { image, error in
+                guard let selectedImage = image as? UIImage else { return }
+                DispatchQueue.main.async { [weak self] in
+                    self?.profileImageView.image =  selectedImage
+                }
+            }
```
P3. 이건 논의를 해봐야 할 것 같은 부분인데요, 메인스레드로 보내는 부분도 Swift Concurrency로 쓰는 건 어떠신가요?? 
``` swift
Task { @MainActor [weak self] in 
    // ,,, 중략 
}
```
요런 느낌입니다!

찾아보니 loadObject가 메인스레드에서 동작하는 게 보장되지 않는다 해서 메인스레드로 보내는 거 아주 좋습니다~~

green-yoon87 replied:
Swift Concurrency의 MainActor를 사용할 수도 있겠군요! 
너무 좋네요!

soletree commented on SniffMeet/SniffMeet/Source/Common/PrimaryButton.swift:
```diff
@@ -36,7 +36,19 @@ final class PrimaryButton: UIButton {
             )
         )
+        
+        let handler: UIButton.ConfigurationUpdateHandler = { button in
+            switch button.state {
+            case .disabled:
+                button.configuration?.baseBackgroundColor = .systemGray3
+            case .normal:
+                button.configuration?.baseBackgroundColor = .mainNavy
+            default:
+                break
+            }
+        }
 
         self.configuration = configuration
+        self.configurationUpdateHandler = handler
```
P5. 오..... 이 부분 완전 신기하네요 소매에 담아갑니다👍

soletree reviewed (commented):
지성님 주말중에도 작업 많이 하셨네요... 

이름 부분 -View로 통일하는 거랑, ProfileSetUpView 네이밍 좋은 것 같습니다. 

P2 부분 하나만 확인 부탁드립니다~~ 고생 많으셨습니다!!

green-yoon87 committed:
[fix] #24 코드 리뷰 적용

- 뷰 작업을 메인스레드로 보내는 부분을 Swift Concurrency를 이용하여 처리
- 텍스트 필드의 텍스트 여부에 따른 버튼 활성화 여부 처리 코드 간결화

kelly-chui commented on SniffMeet/SniffMeet/Source/Auth/CreateProfile/View/ProfileSetupView.swift:
```diff
@@ -0,0 +1,152 @@
+//  Created by 윤지성 on 11/9/24.
+//
+
+import UIKit
+import PhotosUI
```
P4. 둘 다 빌트 인 프레임워크니 알파벳 순서를 지키면 좋을거 같습니다.
```suggestion
import PhotosUI
import UIKit
```

green-yoon87 replied:
넵!

kelly-chui reviewed (approved):
좋아용

PR 특이사항
1. 좋습니다. 지금 다른 PR에 올라가 있는 코드 수정 할게요 (그리고 ViewController랑 View 둘 다 View로 쓰면 되는거죠?)
2. 👍
